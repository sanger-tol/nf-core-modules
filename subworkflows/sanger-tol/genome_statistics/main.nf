include { ASMSTATS            } from '../../../modules/sanger-tol/asmstats/main'
include { BUSCO_BUSCO         } from '../../../modules/nf-core/busco/busco/main'
include { GFASTATS            } from '../../../modules/nf-core/gfastats/main'
include { MERQURYFK_MERQURYFK } from '../../../modules/nf-core/merquryfk/merquryfk/main'

workflow GENOME_STATISTICS {

    take:
    ch_assemblies               // channel: [ val(meta), asm1, asm2 ] - asm2 can be empty
    ch_reads_fastk              // channel: [ val(meta), fastk_hist, [fastk ktabs] ]
    ch_mat_fastk                // channel: [ val(meta), [fastk ktabs] ] - optional
    ch_pat_fastk                // channel: [ val(meta), [fastk ktabs] ] - optional
    val_busco_lineage           // string: busco lineage name
    val_busco_lineage_directory // path: path to local busco lineages directory - optional

    main:
    ch_versions = Channel.empty()

    ch_assemblies_split = ch_assemblies
        | flatMap { meta, asm1, asm2 ->
            def meta_asm1 = meta + [_hap: "hap1"]
            def meta_asm2 = meta + [_hap: "hap2"]
            [ [meta_asm1, asm1], [meta_asm2, asm2] ]
        }
        | filter { meta, asm -> asm }

    //
    // Module: Calculate assembly stats with asmstats
    //
    ASMSTATS(ch_assemblies_split)
    ch_versions = ch_versions.mix(ASMSTATS.out.versions)

    //
    // Module: Calculate assembly stats with gfastats
    //
    GFASTATS(
        ch_assemblies_split, // assembly
        "fasta",             // out_fmt
        [],                  // genome size
        [],                  // target
        [[],[]],             // agp file
        [[],[]],             // include bed
        [[],[]],             // exclude bed
        [[],[]]              // instructions
    )
    ch_versions = ch_versions.mix(GFASTATS.out.versions)

    //
    // Module: Assess assembly using BUSCO_BUSCO
    //
    BUSCO_BUSCO(
        ch_assemblies_split,               // assembly
        "genome",                          // busco mode
        val_busco_lineage,                 // lineage to run BUSCO predictions
        val_busco_lineage_directory ?: [], // busco lineage directory
        [],                                // busco config
        true                               // clean intermediates
    )
    ch_versions = ch_versions.mix(BUSCO_BUSCO.out.versions)

    //
    // Module: assess kmer completeness/QV using MerquryFK
    //
    ch_merquryfk_asm_input = ch_assemblies
        | combine(ch_reads_fastk)
        | map { meta_asm, hap1, hap2, meta_fk, fk_hist, fk_ktabs ->
            [meta_asm, fk_hist, fk_ktabs, hap1, hap2]
        }

    MERQURYFK_MERQURYFK(
        ch_merquryfk_asm_input,
        ch_mat_fastk.ifEmpty([[],[]]),
        ch_pat_fastk.ifEmpty([[],[]])
    )
    ch_versions = ch_versions.mix(MERQURYFK_MERQURYFK.out.versions)
    ch_merquryfk_images = MERQURYFK_MERQURYFK.out.spectra_cn_fl
        | mix(
            MERQURYFK_MERQURYFK.out.spectra_cn_fl,
            MERQURYFK_MERQURYFK.out.spectra_cn_ln,
            MERQURYFK_MERQURYFK.out.spectra_cn_st,
            MERQURYFK_MERQURYFK.out.spectra_asm_fl,
            MERQURYFK_MERQURYFK.out.spectra_cn_ln,
            MERQURYFK_MERQURYFK.out.spectra_cn_st,
            MERQURYFK_MERQURYFK.out.continuity_N,
            MERQURYFK_MERQURYFK.out.block_N,
            MERQURYFK_MERQURYFK.out.block_blob,
            MERQURYFK_MERQURYFK.out.hapmers_blob,
        )

    emit:
    asmstats             = ASMSTATS.out.stats
    gfastats             = GFASTATS.out.assembly_summary
    busco_summary_txt    = BUSCO_BUSCO.out.short_summaries_txt
    busco_summary_json   = BUSCO_BUSCO.out.short_summaries_json
    merqury_qv           = MERQURYFK_MERQURYFK.out.qv
    merqury_completeness = MERQURYFK_MERQURYFK.out.stats
    merqury_phased_stats = MERQURYFK_MERQURYFK.out.phased_block_stats
    merqury_images       = ch_merquryfk_images
    versions             = ch_versions

}
