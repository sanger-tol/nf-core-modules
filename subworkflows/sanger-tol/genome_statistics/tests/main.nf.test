nextflow_workflow {

    name "Test Subworkflow GENOME_STATISTICS"
    script "../main.nf"
    workflow "GENOME_STATISTICS"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/genome_statistics"
    tag "asmstats"
    tag "subworkflows/../../modules/nf-core/busco/busco"
    tag "subworkflows/../../modules/nf-core/gfastats"
    tag "subworkflows/../../modules/nf-core/merquryfk/merquryfk"
    tag "modules/nf-core/fastk/fastk"

    test("meles meles - no trio - no fastk") {

        config "./no_fastk.config"

        setup {
            nfcoreInitialise("${launchDir}/library/")
            nfcoreInstall(
                "${launchDir}/library/",
                [
                    "busco/busco",
                    "gfastats",
                    "merquryfk/merquryfk",
                ]
            )
            nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
        }

        when {

            workflow {
                """
                input[0] = Channel.of([
                    [ id: "test"],
                    file(params.modules_testdata_base_path + 'Ceramica_pisi/assembly/release/ilCerPisi1.1/insdc/GCA_963859965.1.fasta.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Ceramica_pisi/assembly/release/ilCerPisi1.1/insdc/GCA_963859975.1_alt.fasta.gz', checkIfExists: true)
                ])
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = "lepidoptera_odb12"
                input[5] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success               },
                { assert snapshot(workflow.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - no trio - fastk") {

        config "./fastk.config"

        setup {
            nfcoreInitialise("${launchDir}/library/")
            nfcoreInstall(
                "${launchDir}/library/",
                [
                    "busco/busco",
                    "gfastats",
                    "merquryfk/merquryfk",
                    "fastk/fastk",
                ]
            )
            nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")

            run("FASTK_FASTK") {
                script "../../../../modules/nf-core/fastk/fastk/main.nf"
                process {
                    """
                    input[0] = [
                        [ id: "test" ],
                        file(params.modules_testdata_base_path + 'Ceramica_pisi/genomic_data/ilCerPisi1/pacbio/m64097e_230309_154741.ccs.bc1012_BAK8A_OA--bc1012_BAK8A_OA.subsampled.bam', checkIfExists: true)
                    ]
                    """
                }
            }

        }

        when {

            params {
                fastk_args = "-t"
            }

            workflow {
                """
                input[0] = Channel.of([
                    [ id: "test"],
                    file(params.modules_testdata_base_path + 'Ceramica_pisi/assembly/release/ilCerPisi1.1/insdc/GCA_963859965.1.fasta.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Ceramica_pisi/assembly/release/ilCerPisi1.1/insdc/GCA_963859975.1_alt.fasta.gz', checkIfExists: true)
                ])
                input[1] = FASTK_FASTK.out.hist.join(FASTK_FASTK.out.ktab)
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = "lepidoptera_odb12"
                input[5] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success               },
                { assert snapshot(workflow.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

}
