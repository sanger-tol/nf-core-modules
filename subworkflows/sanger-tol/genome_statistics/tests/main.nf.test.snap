{
    "meles meles - no trio - no fastk": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "_hap": "hap1"
                    },
                    "GCA_963859965.1.fasta.gz.stats:md5,27e0c97332ed50fbae78572581db1b56"
                ],
                [
                    {
                        "id": "test",
                        "_hap": "hap2"
                    },
                    "GCA_963859975.1_alt.fasta.gz.stats:md5,bf5abf082c028eee5fbae15da09d9100"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "_hap": "hap1"
                    },
                    "test.assembly_summary:md5,03bd6448d99f284cee9224e932c26336"
                ],
                [
                    {
                        "id": "test",
                        "_hap": "hap2"
                    },
                    "test.assembly_summary:md5,36a8cadec7d7dfd1ec7a922261cf95be"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,09622cb9d9239872f0aa737cff7a2d5f",
                "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36",
                "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-09-01T18:57:36.408961885"
    },
    "meles meles - no trio - fastk": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "_hap": "hap1"
                    },
                    "GCA_963859965.1.fasta.gz.stats:md5,27e0c97332ed50fbae78572581db1b56"
                ],
                [
                    {
                        "id": "test",
                        "_hap": "hap2"
                    },
                    "GCA_963859975.1_alt.fasta.gz.stats:md5,bf5abf082c028eee5fbae15da09d9100"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "_hap": "hap1"
                    },
                    "test.assembly_summary:md5,03bd6448d99f284cee9224e932c26336"
                ],
                [
                    {
                        "id": "test",
                        "_hap": "hap2"
                    },
                    "test.assembly_summary:md5,36a8cadec7d7dfd1ec7a922261cf95be"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.qv:md5,2ad445619abaea809f86268872c1bdae"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.completeness.stats:md5,4499c462a6cae00678eefdeb898e8513"
                ]
            ],
            [
                
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859965.1.spectra-cn.fl.png:md5,c7bd0b65f7fe2a278d8e8342cf24dc45"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859965.1.spectra-cn.ln.png:md5,9e105e1a4bd8b97bc7a0c17075053f61"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859965.1.spectra-cn.st.png:md5,fded9bf30138acdf0d94e542c2f950b8"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859975.1_alt.spectra-cn.fl.png:md5,faa973397243a56ca290203fbcf2ca62"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859975.1_alt.spectra-cn.ln.png:md5,82d5022abd9140e46d0e32e0804d1807"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.GCA_963859975.1_alt.spectra-cn.st.png:md5,e87c4c5ade1e82f341d491660496ae7e"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.spectra-asm.fl.png:md5,4a805b352b0c3d58a260da186b464c0d"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.spectra-asm.ln.png:md5,1da92137877b50f5aa8a0b6758cfa533"
                ],
                [
                    {
                        "id": "test"
                    },
                    "test.spectra-asm.st.png:md5,c16709e5d4bbba67cc16ba1af4f1692b"
                ]
            ],
            [
                "versions.yml:md5,09622cb9d9239872f0aa737cff7a2d5f",
                "versions.yml:md5,6ca1cfcfe9de105329f67f268cfdeeeb",
                "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36",
                "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-09-02T09:53:02.711938626"
    },
    "meles meles - no trio - no busco - no fastk": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "_hap": "hap1"
                        },
                        "GCA_963859965.1.fasta.gz.stats:md5,27e0c97332ed50fbae78572581db1b56"
                    ],
                    [
                        {
                            "id": "test",
                            "_hap": "hap2"
                        },
                        "GCA_963859975.1_alt.fasta.gz.stats:md5,bf5abf082c028eee5fbae15da09d9100"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "_hap": "hap1"
                        },
                        "test.assembly_summary:md5,03bd6448d99f284cee9224e932c26336"
                    ],
                    [
                        {
                            "id": "test",
                            "_hap": "hap2"
                        },
                        "test.assembly_summary:md5,36a8cadec7d7dfd1ec7a922261cf95be"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                    "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                    "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36",
                    "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36"
                ],
                "asmstats": [
                    [
                        {
                            "id": "test",
                            "_hap": "hap1"
                        },
                        "GCA_963859965.1.fasta.gz.stats:md5,27e0c97332ed50fbae78572581db1b56"
                    ],
                    [
                        {
                            "id": "test",
                            "_hap": "hap2"
                        },
                        "GCA_963859975.1_alt.fasta.gz.stats:md5,bf5abf082c028eee5fbae15da09d9100"
                    ]
                ],
                "busco_summary_json": [
                    
                ],
                "busco_summary_txt": [
                    
                ],
                "gfastats": [
                    [
                        {
                            "id": "test",
                            "_hap": "hap1"
                        },
                        "test.assembly_summary:md5,03bd6448d99f284cee9224e932c26336"
                    ],
                    [
                        {
                            "id": "test",
                            "_hap": "hap2"
                        },
                        "test.assembly_summary:md5,36a8cadec7d7dfd1ec7a922261cf95be"
                    ]
                ],
                "merqury_completeness": [
                    
                ],
                "merqury_images": [
                    
                ],
                "merqury_phased_stats": [
                    
                ],
                "merqury_qv": [
                    
                ],
                "versions": [
                    "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                    "versions.yml:md5,ae8574670b9067c621698ce70bdfac49",
                    "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36",
                    "versions.yml:md5,e162b266a7bfe9ab0ae9716e929feb36"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-09-01T16:56:02.670048879"
    }
}